### ISSUE TITLE
Doc.to_bytes() throws KeyError on label-names for rule-matched spans

### ISSUE THREAD

**Issue by wejradford** [explosion/spaCy#1011]
I'd like to run rule-based matching to augment the entities recognised by the model, then write the document out for downstream processing, but it looks like the serialiser can't handle labels I've added myself.

It looks as if it's missing the book-keeping in `nlp.vocab.strings`, so doesn't know how to dump these things out. Is there a neater way to do what I'm trying to, or add the book-keeping manually?

Cheers, Will.

## Info about spaCy

* **spaCy version:** 1.7.5
* **Platform:** Darwin-16.5.0-x86_64-i386-64bit
* **Python version:** 3.6.1
* **Installed models:** en

## To replicate

```python
import spacy
from spacy.matcher import Matcher
nlp = spacy.load('en')
doc = nlp('The cat sat on the mat in Scotland.')
print('Entities\t{}'.format(doc.ents))

print('Dumping')
as_bytes = doc.to_bytes()

matcher = Matcher(nlp.vocab)
matcher.add_entity('ANIMAL')
matcher.add_pattern('ANIMAL', [{'ORTH': 'cat'}], label='ANIMAL')
matches = matcher(doc)
matches = list((None, label_id, start, end)
           for entity_key, label_id, start, end in matches)

print('Matches\t{}'.format(matches))

doc.ents = matches + [(None, e.label, e.start, e.end) for e in doc.ents]
print('Entities + matches\t{}'.format(doc.ents))

print('Dumping')
as_bytes = doc.to_bytes()
```

This outputs:
```
Entities	(Scotland,)
Dumping
Matches	[(None, 63489, 1, 2)]
Entities + matches	(cat, Scotland)
Dumping
Traceback (most recent call last):
  File "replication.py", line 23, in <module>
    as_bytes = doc.to_bytes()
  File "spacy/tokens/doc.pyx", line 607, in spacy.tokens.doc.Doc.to_bytes (spacy/tokens/doc.cpp:12152)
  File "spacy/serialize/packer.pyx", line 112, in spacy.serialize.packer.Packer.pack (spacy/serialize/packer.cpp:5770)
  File "spacy/serialize/huffman.pyx", line 63, in spacy.serialize.huffman.HuffmanCodec.encode (spacy/serialize/huffman.cpp:2554)
KeyError: 63489
```

**Comment by honnibal** [comment]
Thanks!

The general-purpose solution would be to persist and reload the `Vocab`, including the `StringStore`, in order to save/load the document. This is actually the workflow I imagined when I designed the serializer. However, it's not entirely convenient.

The current plan is to replace the existing serializer with a much simpler system that stores the document as a `(text, array)` tuple, with the annotations described in the array. However, you're totally right that this will still raise the problem of these missing strings. I'm not sure what to do about this.

**Comment by wejradford** [comment]
Cool. I could imagine some kind of overflow string store in the document
itself. Most of the time this would be fine and it means you don't need to
block on saving a model to serialise these things out. Although you might
find that the same extra string would get different IDs in different docs 😒

On 24 Apr 2017 8:19 p.m., "Matthew Honnibal" <notifications@github.com>
wrote:

Thanks!

The general-purpose solution would be to persist and reload the Vocab,
including the StringStore, in order to save/load the document. This is
actually the workflow I imagined when I designed the serializer. However,
it's not entirely convenient.

The current plan is to replace the existing serializer with a much simpler
system that stores the document as a (text, array) tuple, with the
annotations described in the array. However, you're totally right that this
will still raise the problem of these missing strings. I'm not sure what to
do about this.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/explosion/spaCy/issues/1011#issuecomment-296607272>,
or mute
the thread
